File: packages/sdk/src/segments.ts
/** How a text splits into SMS parts. An estimate: the phone does the real split. */
export interface SegmentInfo {
  /** `GSM-7` when every character is in the GSM 03.38 alphabet, otherwise `UCS-2`. */
  encoding: "GSM-7" | "UCS-2";
  /** Length in encoding units: septets for GSM-7, UTF-16 code units for UCS-2. */
  length: number;
  /** Number of SMS parts the carrier will count. 0 for an empty text. */
  segments: number;
  /** Units per part at this length: 160 or 153 for GSM-7, 70 or 67 for UCS-2. */
  per_segment: number;
  /** Units that still fit before another part is needed. */
  remaining: number;
}

// GSM 03.38 basic character set. Each costs one septet.
const GSM_BASIC = new Set(
  "@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞÆæßÉ !\"#¤%&'()*+,-./0123456789:;<=>?¡ABCDEFGHIJKLMNOPQRSTUVWXYZÄÖÑÜ§¿abcdefghijklmnopqrstuvwxyzäöñüà",
);

// GSM 03.38 extension table. Each costs two septets (escape plus character).
const GSM_EXTENDED = new Set("\f^{}\\[~]|€");

/**
 * Estimates how many SMS parts `text` needs and which encoding applies.
 * Useful for a counter under a text box or for cost estimates.
 */
export function countSegments(text: string): SegmentInfo {
  let septets = 0;
  let gsm = true;
  for (const ch of text) {
    if (GSM_BASIC.has(ch)) {
      septets += 1;
    } else if (GSM_EXTENDED.has(ch)) {
      septets += 2;
    } else {
      gsm = false;
      break;
    }
  }
  const length = gsm ? septets : text.length;
  const single = gsm ? 160 : 70;
  const multi = gsm ? 153 : 67;
  const per_segment = length <= single ? single : multi;
  const segments = length === 0 ? 0 : Math.ceil(length / per_segment);
  const remaining = segments === 0 ? single : segments * per_segment - length;
  return { encoding: gsm ? "GSM-7" : "UCS-2", length, segments, per_segment, remaining };
}

File: api/internal/sms/segments.go
// Package sms knows how a text is carried on the air: which alphabet it
// needs and how many segments the radio splits it into. Android meters an
// app's outgoing texts by segment, so the count is what pacing is about, and
// it is what the estimate of when a send will be done rests on.
package sms

import "unicode/utf16"

// The GSM 03.38 default alphabet. Each of these is one septet.
const gsmBasic = "@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞÆæßÉ !\"#¤%&'()*+,-./0123456789:;<=>?¡ABCDEFGHIJKLMNOPQRSTUVWXYZÄÖÑÜ§¿abcdefghijklmnopqrstuvwxyzäöñüà"

// The extension table. Each of these costs two septets: an escape and the
// character.
const gsmExtension = "\f^{}\\[~]|€"

const (
	gsmSingle = 160 // septets that fit in one segment
	gsmMulti  = 153 // septets per segment once a concatenation header is needed
	ucsSingle = 70  // UTF-16 units that fit in one segment
	ucsMulti  = 67
)

var (
	basic     = runeSet(gsmBasic)
	extension = runeSet(gsmExtension)
)

func runeSet(s string) map[rune]struct{} {
	m := make(map[rune]struct{}, len(s))
	for _, r := range s {
		m[r] = struct{}{}
	}
	return m
}

// Septets returns how many GSM-7 septets the text needs, and false when it
// holds a character the alphabet cannot carry, which forces UCS-2 for the
// whole text.
func Septets(body string) (int, bool) {
	n := 0
	for _, r := range body {
		switch {
		case isBasic(r):
			n++
		case isExtension(r):
			n += 2
		default:
			return 0, false
		}
	}
	return n, true
}

func isBasic(r rune) bool {
	_, ok := basic[r]
	return ok
}

func isExtension(r rune) bool {
	_, ok := extension[r]
	return ok
}

// Unicode reports whether the text needs UCS-2, which halves what a segment
// holds. Turkish dotless i, ş and ğ are not in the GSM alphabet, so most
// Turkish text does.
func Unicode(body string) bool {
	_, gsm := Septets(body)
	return !gsm
}

// Segments is how many segments the text takes on the air: one while it
// fits, otherwise however many concatenated parts it needs. An empty text
// takes none.
func Segments(body string) int {
	if body == "" {
		return 0
	}
	units, single, multi := 0, ucsSingle, ucsMulti
	if septets, gsm := Septets(body); gsm {
		units, single, multi = septets, gsmSingle, gsmMulti
	} else {
		for _, r := range body {
			units += utf16.RuneLen(r)
		}
	}
	if units <= single {
		return 1
	}
	return (units + multi - 1) / multi
}
